snuba/util.py

```diff
@@ -438,84 +427,70 @@ def raw_query(request: Request, sql, client, timer, stats=None):
-                                # But the query_id will let us know if they aren't
-                                query_id=query_id if use_deduper else None,
-                                with_totals=request.query.get_body().get('totals', False),
-                            )
-                            status = 200
-
-                            logger.debug(sql)
-                            timer.mark('execute')
-                            stats.update({
-                                'result_rows': len(result['data']),
-                                'result_cols': len(result['meta']),
-                            })
-
-                            if use_cache:
-                                state.set_result(query_id, result)
-                                timer.mark('cache_set')
-
-                        except BaseException as ex:
-                            error = str(ex)
-                            status = 500
-                            logger.exception("Error running query: %s\n%s", sql, error)
-                            if isinstance(ex, ClickHouseError):
-                                result = {'error': {
-                                    'type': 'clickhouse',
-                                    'code': ex.code,
-                                    'message': error,
-                                }}
-                            else:
-                                result = {'error': {
-                                    'type': 'unknown',
-                                    'message': error,
-                                }}
-
-                    else:
-                        status = 429
-                        Reason = namedtuple('reason', 'scope name val limit')
-                        reasons = [
-                            Reason('global', 'concurrent', g_concurr, gcl),
-                            Reason('global', 'per-second', g_rate, grl),
-                            Reason('project', 'concurrent', p_concurr, pcl),
-                            Reason('project', 'per-second', p_rate, prl)
-                        ]
-                        reason = next((r for r in reasons if r.limit is not None and r.val > r.limit), None)
-                        result = {'error': {
-                            'type': 'ratelimit',
-                            'message': 'rate limit exceeded',
-                            'detail': reason and '{r.scope} {r.name} of {r.val:.0f} exceeds limit of {r.limit:.0f}'.format(r=reason)
-                        }}
+            with RateLimitAggregator(rate_limits) as (error, rate_limit_stats):
+                stats.update(rate_limit_stats)
+                if not error:
+                    # Experiment, reduce max threads by 1 for each extra concurrent query
+                    # that a project has running beyond the first one
+                    if 'max_threads' in query_settings and \
+                            'project_concurrent' in rate_limit_stats \
```

Comment:
The pesky `project` reappears. 😞 

The alternative, however, is to provide the `query_settings` when we create the rate limits, which are all the way in api.py and that doesn't feel like a good separation of concerns.

Reply:
First I would verify whether this experiment is still active in production. Maybe this code is not relevant anymore.

Reply:
It is, and I would prefer to keep it that way.

It is one of our biggest pressure/performance scalars.

If the max_threads + per_project concurrent stuff is too hard to weave into this work, then we can attempt to ditch it, but the reduction of used threads while concurrency increases is a nice back-pressure dial.

Reply:
ok, then if we keep it, I would advise we keep this logic here, but we make RateLimitStats a bit more manageable to prevent us from changing the name of the extension and seeing this  breaking silently.
Think about trying to find "project_concurrent" in the codebase, you won't find it because it is a string built on the fly.
Why not turning that Mapping into a class with a getter that takes the rate limiter name and returns the stats. And then defining the rate limiter name in a constant.